### PR TITLE
Support qwen3-4B,qwen3.5-9B&35B training on P800

### PR DESCRIPTION
**Pull request by yeqinghe609**
## What

Add end-to-end Kunlunxin P800 (XPU) training support for Relax, including ready-to-run scripts and a complete user guide. Verified accuracy parity with H20 on Qwen3-4B, Qwen3.5-9B, and Qwen3.5-35B-A3B (DAPO).

## Why

Enable Relax users to run RL training (GRPO / DAPO) on P800 with the same algorithms, recipes, and convergence behavior as on H20, without having to maintain a separate fork.

## How

- **Accuracy alignment with H20**
  - Validated reward / KL / response-length curves on Qwen3-4B, Qwen3.5-9B, Qwen3.5-35B-A3B against H20 baselines — curves overlap within run-to-run variance (see screenshots in `docker/xpu-prevision-images/`).
  - Stabilized the Megatron-Bridge weight-export path used during rollout weight sync so that policy weights pushed to SGLang are byte-identical to the trainer state.
- **Model & script support**
  - New launch scripts under `scripts/training/text/`:
    - `run-qwen3-4B-8xP800.sh` (single node, 8 cards)
    - `run-qwen35-9B-8xP800.sh` (single node, 8 cards)
    - `run-qwen35-35B-A3B-16xP800.sh` (two nodes, 16 cards, MoE)
  - Tuned `PERF_ARGS`, `OPTIMIZER_ARGS` (CPU optimizer offload), and `SGLANG_ARGS` per model size; MoE-specific flags for 35B-A3B.
- **Self-contained patch + image**
  - `docker/patch/xpu_patch/relax.patch` (based on `2100c156`) and `docker/patch/xpu_patch/megatron.patch` (based on Megatron-LM `85bced0ae`) capture all framework-side adaptations needed on top of mainline.
  - Container image published as `iregistry.baidu-int.com/xpu/xrelax_torch29_ubuntu2204_xsgl0510_dev:20260610_12`, with an offline tarball mirror documented in `docker/xpu-training.md`.

## Testing

- Trained Qwen3-4B / Qwen3.5-9B (DAPO) for >200 steps on a single P800 node; reward / KL / response-length match H20 reference runs.
- Trained Qwen3.5-35B-A3B (DAPO) for >200 steps across 2× P800 nodes (16 cards) in both sync (colocate) and async modes; convergence matches H20.
- `pre-commit run --all-files` passes.
- Unit tests in `tests/` unaffected; XPU-specific runs gated behind hardware availability.

- [x] `pre-commit run --all-files` passes
- [ ] Tests pass (`pytest tests/`) — XPU-only runs require P800 hardware
- [ ] New tests added (if applicable)
- [x] Documentation updated (`docker/xpu-training.md`)

## Type of Change

- [x] New feature (non-breaking change that adds functionality)
- [x] Documentation update

## Usage

```bash
# 1. Pull (or load offline) the published XPU image

# 2. Inside the container, place Relax at /workspace/Relax and apply patches
cd /workspace/Relax        && git checkout 2100c156 && git apply docker/patch/xpu_patch/relax.patch
cd /workspace/Megatron-LM  && git checkout 85bced0ae && git apply /workspace/Relax/docker/patch/xpu_patch/megatron.patch

# 3. Launch training
# Single node (4B / 9B):
bash scripts/training/text/run-qwen3-4B-8xP800.sh
bash scripts/training/text/run-qwen35-9B-8xP800.sh

# Two nodes (35B-A3B):
export HEAD_IP=<head-ip> WORKER_NODES=<worker-ip>
bash scripts/training/text/run-qwen35-35B-A3B-16xP800.sh

### CODE REVIEW

File: docker/xpu-training.md
Comment (Aurelius84) [inline docker/xpu-training.md:1]: 1. xpu-prevision-images 目录移动到顶层 assets 目录，统一放image文件
2. image 文件命名需要规范，类似raw_reward.png，不同模型或recipe建议组织成子目录

File: docker/xpu_patch/relax.patch [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+                 else:
+                     raise
+ 
++        self._offload_train_enabled = bool(args.offload_train)
++
+         if role == "critic":
+             self.args.load = self.args.critic_load
+             self.args.save = self.args.critic_save
+@@ -254,13 +257,12 @@
+     def sleep(self) -> None:
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:21]: sleep 和 wakeup 函数可以抽离成单独的一个类? 不同硬件若有需要，根据device 特化不同的实现。在actor.py里只是转发调用此类对应的函数即可？cc @NINGBENZHE @Yangruipis

File: docker/xpu_patch/relax.patch [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+-            torch_memory_saver.resume()
++        if self._offload_train_enabled:
++            self._reload_train_to_gpu()
+ 
+         clear_memory()
+-        reload_process_groups(timeout_minutes=self.args.distributed_timeout_minutes)
+         print_memory("after wake_up model")
+ 
++    # ------------------------------------------------------------------
++    # Explicit CPU offload for XPU (torch_memory_saver VMM pause is unsafe here).
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:51]: 如下XPU 特化的函数应该内聚要单独类似 xpu_utils.py 文件里，不建议放到actor.py里。新增文件直接放代码仓

File: docker/xpu_patch/relax.patch [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+ 
++from hydrax.Hydra import TensorState
+ 
+ logging.getLogger("megatron").setLevel(logging.WARNING)
+ 
+@@ -133,6 +134,8 @@
+                 else:
+                     raise
+ 
++        self._offload_train_enabled = bool(args.offload_train)
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:15]: 这个其实可以复用self.args.offload_train?

File: docker/xpu_patch/relax.patch [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+-    quant_weight_ue8m0 = None
+-    transform_scale_ue8m0 = None
+-    should_deepgemm_weight_requant_ue8m0 = None
++# try:
++#     from sglang.srt.layers.quantization.fp8_utils import quant_weight_ue8m0, transform_scale_ue8m0
++#     from sglang.srt.model_loader.utils import should_deepgemm_weight_requant_ue8m0
++# except ImportError:
++quant_weight_ue8m0 = None
++transform_scale_ue8m0 = None
++should_deepgemm_weight_requant_ue8m0 = None
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:265]: 建议放到repo代码里，做好分支管理

File: docker/xpu_patch/relax.patch [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
++import torch
++
+ from relax.utils import megatron_bridge_utils
+ from relax.utils.misc import chunk_named_params_by_size
+ 
+@@ -20,6 +22,11 @@ class HfWeightIteratorBridge(HfWeightIteratorBase):
+         self._bridge = AutoBridge.from_hf_pretrained(self.args.hf_checkpoint, trust_remote_code=True)
+ 
+     def get_hf_weight_chunks(self, megatron_local_weights):
++        # Drain pending async H2D/D2H ops (e.g. from train-step backup or
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:316]: 这个大函数的改动是硬件无关的「功能扩展」，还是适配P800的特定逻辑？

File: docker/xpu_patch/relax.patch [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+         if prev_refs:
+             ray.get(prev_refs)
+--- a/relax/components/actor.py	2026-05-08 19:17:00.172660820 +0800
++++ b/relax/components/actor.py	2026-06-08 11:21:41.462566333 +0800
+@@ -81,6 +81,8 @@
+         # In fully_async mode, weight_updater is not created and weights are synced via DCS
+         if not self.config.fully_async:
+             self.actor_model.update_weights()
++            if getattr(self.config, "check_weight_update_equal", False):
++                ray.get(self.rollout_manager.check_weights.remote(action="compare"))
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:401]: 这个是生产环境上必要的代码逻辑吗？

File: docker/xpu_patch/relax.patch [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+--- a/relax/distributed/ray/actor_group.py	2026-05-08 19:17:00.284658431 +0800
++++ b/relax/distributed/ray/actor_group.py	2026-06-08 11:21:41.466566243 +0800
+@@ -71,8 +71,8 @@
+             assert os.path.exists(dynlib_path), f"LD_PRELOAD so file {dynlib_path} does not exist."
+ 
+             env_vars["LD_PRELOAD"] = dynlib_path
+-            env_vars["TMS_INIT_ENABLE"] = "1"
+-            env_vars["TMS_INIT_ENABLE_CPU_BACKUP"] = "1"
++            # env_vars["TMS_INIT_ENABLE"] = "1"
++            # env_vars["TMS_INIT_ENABLE_CPU_BACKUP"] = "1"
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:414]: TMS 操作看是否可以在XPU 特定的package逻辑里做兼容？

File: docker/xpu_patch/relax.patch [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+ from relax.utils.logging_utils import get_logger
++from relax.utils.tracking_utils import init_tracking
+ from relax.utils.utils import post_process_env
+ 
+ 
+@@ -59,6 +60,7 @@
+         runtime_env = yaml.safe_load(file)
+ 
+     runtime_env = post_process_env(args, runtime_env)
++    init_tracking(args, primary=True)
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:432]: 这个是针对relax 的bug fix？

File: docker/xpu_patch/relax.patch [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+ 
+ 
+ def init_tracking(args, primary: bool = True, **kwargs):
+@@ -31,7 +33,13 @@
+ 
+     # Fall back to direct logging if metrics service is not used or not available
+     if args.use_wandb:
+-        wandb.log(metrics)
++        if wandb.run is None:
++            logger.warning(
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:458]: 这个是针对relax 的bug fix？

File: docker/xpu_patch/relax.patch [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+ 
+ 
+--- a/scripts/entrypoint/local.sh	2026-06-10 15:46:08.585492936 +0800
++++ b/scripts/entrypoint/local.sh	2026-06-10 16:00:00.000000000 +0800
+@@ -39,9 +39,9 @@
+     return 0 2>/dev/null || exit 0
+ fi
+ 
+-set -eo pipefail
++# set -eo pipefail
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:496]: P800 上打开这个会有什么影响？

File: docker/xpu_patch/relax.patch [hunk 11]
```diff
@@ -0,0 +1,553 @@
...
+         metrics_except_step = {k: v for k, v in metrics.items() if k != step_key}
+--- a/relax/utils/training/tensor_backper.py	2026-05-08 19:17:00.792647594 +0800
++++ b/relax/utils/training/tensor_backper.py	2026-06-08 11:21:41.474566065 +0800
+@@ -59,8 +59,8 @@
+         backup_dict = self._backups[tag]
+         for name, param in self._source_getter():
+             if name not in backup_dict:
+-                backup_dict[name] = torch.empty_like(param, device=torch.device("cpu"), pin_memory=True)
+-            backup_dict[name].copy_(param.detach(), non_blocking=True)
++                backup_dict[name] = torch.empty_like(param, device=torch.device("cpu"))
```
Comment (Aurelius84) [inline docker/xpu_patch/relax.patch:475]: 后续P800上会支持开启pin_memory吗？

File: docker/xpu_patch/relax.patch (hunk 7 above)
Comment (yeqinghe609) [reply docker/xpu_patch/relax.patch:414]: 嗯嗯 好的

File: docker/xpu_patch/relax.patch (hunk 8 above)
Comment (yeqinghe609) [reply docker/xpu_patch/relax.patch:432]: 是的 这个在使用runtime_env_json时有点问题，我提个对应pr

File: docker/xpu_patch/relax.patch (hunk 11 above)
Comment (yeqinghe609) [reply docker/xpu_patch/relax.patch:475]: p800支持pin_memory.这里开启之前在verl上p800和h20都出现过问题，这个问题我们再追下
